src/common/libjob/jobspec1.c

```diff
@@ -342,35 +509,67 @@ int flux_jobspec1_check (flux_jobspec1_t *jobspec, flux_jobspec1_error_t *error)
 
 int flux_jobspec1_unsetenv (flux_jobspec1_t *jobspec, const char *name)
 {
-    json_t *environment;
+    char *path;
 
     if (!jobspec || !name) {
         errno = EINVAL;
         return -1;
     }
-    if (!(environment = specutil_attr_get (jobspec->obj,
-                                           "attributes.system.environment"))) {
+    if (asprintf (&path, "system.environment.%s", name) < 0)
+        return -1;
+    if (flux_jobspec1_attr_del (jobspec, path) < 0) {
+        ERRNO_SAFE_WRAP (free, path);
         return -1;
     }
-    return specutil_env_unset (environment, name);
+    return 0;
 }
 
 int flux_jobspec1_setenv (flux_jobspec1_t *jobspec,
                            const char *name,
                            const char *value,
                            int overwrite)
 {
-    json_t *environment;
+    char *path;
+    const char *s;
 
     if (!jobspec || !name || !value) {
         errno = EINVAL;
         return -1;
     }
-    if (!(environment = specutil_attr_get (jobspec->obj,
-                                           "attributes.system.environment"))) {
+    if (asprintf (&path, "system.environment.%s", name) < 0)
+        return -1;
+    if (!overwrite) {
+        if (flux_jobspec1_attr_unpack (jobspec, path, "s", &s) == 0) {
+            free (path);
+            return 0;
+        }
+    }
+    if (flux_jobspec1_attr_pack (jobspec, path, "s", value) < 0) {
+        ERRNO_SAFE_WRAP (free, path);
+        return -1;
+    }
+    free (path);
+    return 0;
+}
+
+static int jobspec1_putenv (flux_jobspec1_t *jobspec, const char *entry)
+{
+    char *name;
+    char *value;
+    int rc = -1;
+
+    if (!(name = strdup (entry)))
         return -1;
+    if ((value = strchr (name, '=')))
+        *value++ = '\0';
+    if (!value || value == name + 1) {
+        errno = EINVAL;
+        goto done;
```

Comment:
perhaps this is obvious and I'm just not seeing it, but what is the purpose of the `value == name + 1` corner case check?  Is for when the user doesn't specify the environment variable name? i.e. `=value`?  I think `name[0] == '\0'` might be more clear?

Also, should a user be able to specify an environment variable without a value?

Reply:
That was a weird way to write it, agreed :-/  Will fix.

> Also, should a user be able to specify an environment variable without a value?

I'll add a test that a user can set an environment var whose value is an empty string since `putenv ("name=")` or `setenv ("name", "")` seem to be allowed.